### DISCUSSION TITLE
Getting request and return types of API endpoints when using typescript

### DISCUSSION THREAD

**Discussion by NixBiks** [prisma/prisma#9926] [Q&A]
Is it possible to easily get the request and return types of API endpoint when using typescript?

Usually I run some typescript codegen based on OpenAPI schema created from my backend but maybe there is a better way when backend and frontend are in the exact same codebase?

I wonder if anyone came up with a good pattern

**Comment by ryands17** [comment]
@mr-bjerre 👋 

What you're looking for would be this: https://trpc.io/

This would provide end-to-end type safety if your backend and frontend are in the same codebase 🙂

**Reply by NixBiks** [reply]
@ryands17 looks promising indeed. I was hoping for something a little less intrusive though - this changes the way of writing the api completely. 

I was hoping for something that could exploit typescript to produce openapi.json which then can be used to produce a typesafe client. Or even short circuit and produce the client directly from a typescripted NextJS api.

**Reply by ryands17** [reply]
Maybe you could try something like this then: https://www.npmjs.com/package/openapi-typescript

**Reply by NixBiks** [reply]
Yeah so that's the step from openapi to ts client. I'm trying to see if there's a way to come from nextjs typescript to openapi.  Thereby from api to ts client

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
